src/render/SDL_palette_sw.c

```diff
@@ -0,0 +1,165 @@
+        SDL_SW_DestroyPaletteTexture(swdata);
+        return NULL;
+    }
+
+    SDL_SetSurfacePalette(swdata->surface, swdata->palette);
+
+    /* We're all done.. */
+    return (swdata);
+}
+
+int
+SDL_SW_UpdatePaletteTexture(SDL_SW_PaletteTexture* swdata, const SDL_Rect* rect,
+                            const void* pixels, int pitch,
+                            const SDL_Color* colors, int firstcolor, int ncolors)
+{
+    const Uint8 *src;
+    Uint8 *dst;
+    int row;
+    size_t length;
+
+    if (pixels) {
+        if (SDL_LockSurface(swdata->surface) < 0)
+            return -1;
+
+        src = pixels;
+        dst = (Uint8*)swdata->surface->pixels + rect->y * swdata->surface->pitch + rect->x;
+        length = rect->w;
+        for (row = 0; row < rect->h; ++row) {
+            SDL_memcpy(dst, src, length);
+            src += pitch;
+            dst += swdata->surface->pitch;
+        }
+
+        SDL_UnlockSurface(swdata->surface);
+    }
+    if (colors) {
+        SDL_SetPaletteColors(swdata->palette, colors, firstcolor, ncolors);
+    }
+    return 0;
+}
+
+int
+SDL_SW_LockPaletteTexture(SDL_SW_PaletteTexture * swdata, const SDL_Rect * rect,
+                          void **pixels, int *pitch)
+{
+    if (SDL_LockSurface(swdata->surface) < 0)
+        return -1;
+
+    if (rect) {
+        *pixels = (Uint8 *)swdata->surface->pixels + rect->y * swdata->surface->pitch + rect->x;
+    } else {
+        *pixels = (Uint8 *)swdata->surface->pixels;
+    }
+    *pitch = swdata->surface->pitch;
+    return 0;
+}
+
+void
+SDL_SW_UnlockPaletteTexture(SDL_SW_PaletteTexture * swdata)
+{
+    SDL_UnlockSurface(swdata->surface);
+}
+
+int
+SDL_SW_CopyPaletteToRGB(SDL_SW_PaletteTexture * swdata, const SDL_Rect * srcrect,
+                        Uint32 target_format, int w, int h, void *pixels,
+                        int pitch)
+{
+    SDL_Rect rect = *srcrect;
+
+    /* Make sure we're set up to display in the desired format */
+    if (swdata->display && target_format != swdata->display->format->format) {
+        SDL_FreeSurface(swdata->display);
+        swdata->display = NULL;
+    }
+
+    if (swdata->display) {
+        swdata->display->w = w;
+        swdata->display->h = h;
+        swdata->display->pixels = pixels;
+        swdata->display->pitch = pitch;
+    } else {
+        swdata->display =
+            SDL_CreateRGBSurfaceWithFormatFrom(pixels, w, h, SDL_BITSPERPIXEL(target_format),
+                                               pitch, target_format);
+        if (!swdata->display) {
+            return (-1);
+        }
+    }
+
+    return SDL_BlitSurface(swdata->surface, &rect, swdata->display, NULL);
+}
+
+void
+SDL_SW_DestroyPaletteTexture(SDL_SW_PaletteTexture * swdata)
+{
+    if (swdata) {
+        SDL_FreeSurface(swdata->display);
+        SDL_FreeSurface(swdata->surface);
+        SDL_FreePalette(swdata->palette);
+        SDL_free(swdata);
+    }
+}
+
+/* vi: set ts=4 sw=4 expandtab: */
```

Comment:
Despite of fact this line is present in other files this is an editor specific thing which shouldn't exist in the generic source code. This is a developer responsibility to adjust editors per their needs.